server/aesdsocket.c

```diff
@@ -53,39 +78,183 @@ const char* errs[] = {
 	[SRC_WRITE] = "error writing to client socket",
 	[SRC_FTRUNCATE] = "error extending output file",
 	[SRC_EINVAL] = "parsing command line arguments",
-	[SRC_DUP] = "error redirecting stream"
+	[SRC_DUP] = "error redirecting stream",
+	[SRC_PTHCR] = "error creating client thread",
+	[SRC_PTHJN] = "error joining client thread",
+	[SRC_STRFTIME] = "error formatting time",
+	[SRC_PTHATTR] = "error setting thread attributes"
 };
 
-// cleanup handler
+// main cleanup handler
 __attribute__((noreturn))
-static void cleanup (cleanup_src src) {
+static void cleanup_errno (cleanup_src src, int e) {
+	// cleanup threads
+	for (struct node* n = thread_ll; n; n = n->next)
+		pthread_cancel(n->tid);
+	for (struct node* n = thread_ll; n; n = n->next)
+		pthread_join(n->tid, NULL);
+
 	// record exit cause
 	const char* msg = errs[src];
 	int xstat = 0;
-	if (src == SRC_INT || src == SRC_TERM) {
+	if (e == 0) {
 		syslog(LOG_INFO, "%s\n", msg);
 	} else {
-		syslog(LOG_ERR, "%s: %m", msg);
+		syslog(LOG_ERR, "%s: %s", msg, strerror(e));
 		xstat = -1;
 	}
 	closelog();
 
 	// clean up state
-	if (csock != -1) close(csock);
+	struct node* p;
+	for (struct node* n = thread_ll; n; n = p) {
+		p = n->next;
+	   	free(n);
+	}	
 	if (asock != -1) close(asock);
 	if (of_mem) munmap(of_mem, of_memsz);
 	if (ofd != -1) close (ofd);
-	if (packet) free(packet);
 	unlink(outpath);
 
 	exit(xstat);
 }
 
-// wrapper to conform to sigaction struct def
+// synchronous main thread cleanup function
+__attribute__((noreturn))
+static void cleanup (cleanup_src src) {
+	cleanup_errno(src, errno);
+}
+
+// async SIGTERM/SIGINT handler
 __attribute__((noreturn))
-static void sigcleanup (int sig) {
+static void sigcleanup (int sig, siginfo_t* info, void* unused) {
 	assert(sig == SIGTERM || sig == SIGINT);
-	cleanup((cleanup_src) sig);
+	if (info->si_code == SI_QUEUE) {
+		int32_t* unpack = (int32_t*) &info->si_ptr; 
+		cleanup_errno((cleanup_src) unpack[0], unpack[1]);
+	} else cleanup_errno((cleanup_src) sig, 0);
+}
+
+// cleanup handler for threads (sends SIGTERM with error info)
+__attribute__((noreturn))
+static void cleanup_thr (cleanup_src src) {
+	int32_t pack[2] = {src, errno};
+	union sigval v = {.sival_ptr = (void*) *(uint64_t*) pack};
+	sigqueue(getpid(), SIGTERM, v);
+	pthread_exit(NULL);
+}
+
+//-----client thread-----
+
+// close socket if cancelled
+static void vclose(void* fdpt) {
+	close(*(int*) fdpt);
+}
+
+static void* client_thread (void* param_v) {
+	// these must be declared before pushing the cleanup handler
+	// because it creates a scope...
+	int s;
+	char* cip;
+	struct client* param = (struct client*) param_v;
+	pthread_cleanup_push(vclose, &param->sock);
+
+	// log accepted connection
+	cip = alloca(INET_ADDRSTRLEN);
+	pthread_mutex_lock(&ntoa_lk);
+	const char* ipstr = inet_ntoa(param->addr);
```

Comment:
@georgehodgkins this works to lock around `inet_ntoa` but you might want to consider [inet_ntop](https://linux.die.net/man/3/inet_ntop) instead so you don't need the extra lock.